### PR TITLE
compatibility with _Pragma

### PR DESCRIPTION
**Pull request by nilput**
I'm not totally sure if this is the correct place to add it, i mean it could be better if it was handled at the preprocessor, this could serve as a temporary hack though, it works in the few test cases i've tried.

i'm not sure about the string allocation, by comparing it to the "asm" statements, it seems no memory is explicitly deallocated?

Best regards.

### CODE REVIEW

**Comment by MatzeB** [comment]
This should better be handled in the preprocessor (it's even described in the C99 standard). So I think you would better try handling it like a function macro. You can probably do something similar to
```
    pp_definition_t *def = add_define_("_Pragma", true);

    symbol_t *const parameter_symbol = symbol_table_insert("directive");
    pp_definition_t *parameter = OALLOCZ(&pp_obstack, pp_definition_t);
    parameter->pos          = builtin_position;
    parameter->symbol       = parameter_symbol;
    parameter->is_parameter = true;
    parameter->is_variadic  = true;

    def->has_parameters = true;
    def->n_parameters   = 1;
    def->parameters     = parameter;
```
and then catch it when expansion is started. I also recommend writing a testcase or two for the firm-testsuite so you can try it behaves as expected when the pragma is put at various places in the program (between functions, as statement, between an expresison, expanded through a #define...)

**Comment by nilput** [comment]
@MatzeB i see, doing what you suggested and adding a new predefined macro works.

about the tests, wouldn't it be a good idea to have a directory for cparser's tests?
they would ideally run after building cparser itself and have input files and expected results in the outputs.
many tests would need only simple shell scripts like:

```
./build/debug/cparser -o "$test_result" -E "$test_input" 2>"$test_stderr"
grep 'Wunknown-pragmas' "$test_stderr" && exit 0
exit 1
```

even results expected out of a compiled input can be reduced to something a shell script can verify.

File: src/driver/predefs.c [hunk 1]
```diff
@@ -453,6 +453,8 @@ void add_predefined_macros(void)
 
 	/* TODO: __CHAR16_TYPE__, __CHAR32_TYPE__ */
 
+    define_special_macro_functions();
```
Comment (MatzeB) [inline src/driver/predefs.c]: I think this needs to be indented with tabs (unless this is a display problem in github)

File: src/parser/preprocessor.c [hunk 2]
```diff
@@ -60,13 +60,14 @@ struct pp_definition_t {
...
-	bool             standard_define : 1;
-	bool             not_specified   : 1;
+    bool             is_expanding           : 1;
+    bool             may_recurse            : 1;
+    bool             has_parameters         : 1;
+    bool             is_parameter           : 1;
+    bool             is_variadic            : 1;
+    bool             standard_define        : 1;
+    bool             not_specified          : 1;
+    bool             is_specially_handled   : 1; /*predefined macro functions that are handled differntly*/
```
Comment (MatzeB) [inline src/parser/preprocessor.c]: - Indent with tabs.
- You could just call the new field `is_pragma`. We can always rename later if we find other cases.

File: src/parser/preprocessor.c [hunk 3]
```diff
@@ -1126,6 +1127,10 @@ static void start_function_macro_expansion(const macro_call_t *call)
 			parameter->not_specified = true;
 		}
 	}
+    if (current_call.macro->is_specially_handled){
+        //TODO: this is done temporarily for _Pragma
```
Comment (MatzeB) [inline src/parser/preprocessor.c]: We tend to use `/*` in cparser and always leave at least 1 space after opening the comment (similar at other places in the patch).

File: src/parser/preprocessor.c [hunk 4]
```diff
@@ -2670,6 +2675,26 @@ void add_define_macro(char const *const name, char const *const macro_arg,
 	def->token_list     = obstack_finish(&pp_obstack);
 }
 
+
+void define_special_macro_functions()
```
Comment (MatzeB) [inline src/parser/preprocessor.c]: - How about `define_pragma_macro()`?
- Use `(void)` (`()` unfortunately means "unknown parameter count", rather than "no parameters" in C)

File: src/parser/preprocessor.c [hunk 5]
```diff
@@ -2670,6 +2675,26 @@ void add_define_macro(char const *const name, char const *const macro_arg,
...
+    pp_definition_t *def = add_define_("_Pragma", true);
+
+    symbol_t *const parameter_symbol = symbol_table_insert("directive");
+    pp_definition_t *parameter = OALLOCZ(&pp_obstack, pp_definition_t);
+    parameter->pos          = builtin_position;
+    parameter->symbol       = parameter_symbol;
+    parameter->is_parameter = true;
+    parameter->is_variadic  = false;
+
+    def->has_parameters = true;
```
Comment (MatzeB) [inline src/parser/preprocessor.c]: align equal signs.

File: src/parser/preprocessor.c [hunk 6]
```diff
@@ -4064,7 +4089,7 @@ try_input:;
 			if (current_call.argument_brace_count > 0) {
 				--current_call.argument_brace_count;
 			} else {
-				assert(kind == ')');
+				assert(current_call.argument_brace_count == 0);
```
Comment (MatzeB) [inline src/parser/preprocessor.c]: Is this change necessary for the _Pragma support or unrelated?

Comment (nilput) [reply src/parser/preprocessor.c]: no it's not!, the assertion is redundant as you can see in the if condition leading to the current branch. i thought this was the original intention from the context.

File: src/driver/predefs.c [hunk 7]
```diff
@@ -453,6 +453,8 @@ void add_predefined_macros(void)
 
 	/* TODO: __CHAR16_TYPE__, __CHAR32_TYPE__ */
 
+    define_pragma_macro();
```
Comment (rofl0r) [inline src/driver/predefs.c]: whitespace still wrong

File: src/parser/preprocessor.c [hunk 8]
```diff
@@ -1126,6 +1127,10 @@ static void start_function_macro_expansion(const macro_call_t *call)
 			parameter->not_specified = true;
 		}
 	}
+    if (current_call.macro->is_pragma){
```
Comment (rofl0r) [inline src/parser/preprocessor.c]: wrong indentation again... maybe you should get a better editor

Comment (nilput) [reply src/parser/preprocessor.c]: sorry i didn't pay attention,
i use vim, it's just that i'm not used to contributing to projects, i should stick to project specific settings next times

File: src/parser/preprocessor.c [hunk 9]
```diff
@@ -1126,6 +1127,10 @@ static void start_function_macro_expansion(const macro_call_t *call)
 			parameter->not_specified = true;
 		}
 	}
+	if (current_call.macro->is_pragma){
+	/*TODO: this is done temporarily for _Pragma*/
+        warningf(WARN_UNKNOWN_PRAGMAS, &pp_token.base.pos, "_Pragma is not currently supported");
```
Comment (rofl0r) [inline src/parser/preprocessor.c]: indentation wrong

**Comment by nilput** [comment]
used git diff this time, but it's too late all spaces have been spotted.

**Comment by rofl0r** [comment]
i think this hasn't been merged so far due to https://github.com/libfirm/cparser/pull/16#discussion_r227626747 . there's no good reason to have unrelated changes in the same commit.
